other/dpll.py

```diff
@@ -0,0 +1,400 @@
+        """
+        clause = "{ "
+        for i in range(0, self.no_of_literals):
+            clause += str(list(self.literals.keys())[i]) + " "
+            if i != self.no_of_literals - 1:
+                clause += ", "
+        clause += "}"
+
+        return clause
+
+    def assign(self, model):
+        """
+        Assign values to literals of the clause as given by model.
+        """
+        i = 0
+        while i < self.no_of_literals:
+            symbol = list(self.literals.keys())[i][:2]
+            if symbol in model.keys():
+                val = model[symbol]
+            else:
+                i += 1
+                continue
+            if val is not None:
+                # Complement assignment if literal is in complemented form
+                if list(self.literals.keys())[i][-1] == "'":
+                    val = not val
+            self.literals[list(self.literals.keys())[i]] = val
+            i += 1
+
+    def evaluate(self, model):
+        """
+        Evaluates the clause with the assignments in model.
+        This has the following steps:
+        1. Return True if both a literal and its complement exist in the clause.
+        2. Return True if a single literal has the assignment True.
+        3. Return None(unable to complete evaluation) if a literal has no assignment.
+        4. Compute disjunction of all values assigned in clause.
+        """
+        for literal in list(self.literals.keys()):
+            if len(literal) == 2:
+                symbol = literal + "'"
+                if symbol in list(self.literals.keys()):
+                    return True
+            else:
+                symbol = literal[:2]
+                if symbol in list(self.literals.keys()):
+                    return True
+
+        self.assign(model)
+        result = False
+        for j in self.literals.values():
+            if j is True:
+                return True
+            elif j is None:
+                return None
+        for j in self.literals.values():
+            result = result or j
+        return result
+
+
+class Formula:
+    """
+    A formula represented in CNF.
+    A formula is a set of clauses.
+    For example,
+        {{A1, A2, A3'}, {A5', A2', A1}} is ((A1 v A2 v A3') and (A5' v A2' v A1))
+
+    Create two clauses and a formula with them
+    >>> c1 = Clause(["A1", "A2'", "A3"])
+    >>> c2 = Clause(["A5'", "A2'", "A1"])
+
+    >>> f = Formula([c1, c2])
+    >>> print(f)
+    { { A1 , A2' , A3 } , { A5' , A2' , A1 } }
+    """
+
+    def __init__(self, clauses):
+        """
+        Represent the number of clauses and the clauses themselves.
+        """
+        self.clauses = [c for c in clauses]
+        self.no_of_clauses = len(self.clauses)
+
+    def __str__(self):
+        """
+        To print a formula as in CNF.
+        Variable formula holds the string representation.
+        """
+        formula = "{ "
+        for i in range(0, self.no_of_clauses):
```

Comment:
```suggestion
        for clause in self.clauses:
```
We should use `str.join()` to create this string.